DependencyInjectionContainer/DIContainer.cs

```diff
@@ -1,35 +1,48 @@
 ﻿using System.Reflection;
+using System.Runtime.CompilerServices;
 using System.Runtime.ExceptionServices;
 using DependencyInjectionContainer.Enums;
 using DependencyInjectionContainer.Exceptions;
 
 namespace DependencyInjectionContainer
 {
-    public sealed class DIContainer
+    public sealed class DIContainer : IDisposable
     {
-        private readonly IEnumerable<Service> registeredServices;
+        internal readonly IEnumerable<Service> registeredServices;
+        ServicesInstanceList servicesInstanceList;
 
         internal DIContainer(IEnumerable<Service> services, DIContainer parent)
         {
+            servicesInstanceList = new ServicesInstanceList();
             registeredServices = services;
             ContainerParent = parent;
+            IsDisposed = false;
         }
 
-        internal DIContainer ContainerParent { get; init; }
+        public bool IsDisposed { get; private set; }
```

Comment:
Let's throw on double dispose so consider can hide this property and check exception it tests on second dispose